divintseg/_core.py

```diff
@@ -422,3 +422,151 @@ def similarity(
+
+    >>> from divintseg import isolation
+    ...
+    ... isolation(df, "S", by="REGION", over="SUBREGION")
+         REGION        S
+    0  Region 1  0.83333
+    1  Region 2      0.1
+
+    Let's look at what this example computed. First, we have
+    to see how likely each person in group S is to see other
+    members of their own group in their subregion. This is as
+    follows:
+
+    +----------+-------------+-----------------------+
+    | Region   | Subregion   | Likelihood of an S    |
+    +==========+=============+=======================+
+    | Region 1 | Subregion A | 100 / (100 + 0) = 1.0 |
+    +----------+-------------+-----------------------+
+    | Region 1 | Subregion B | 50 / (50 + 50) = 0.5  |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion C | 0 / (0 + 100) = 0.0   |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion D | 0 / (0 + 50) = 0.0    |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion E | 10 / (10 + 90) = 0.1  |
+    +----------+-------------+-----------------------+
+
+    Next, we can compute the fraction of all S's in
+    each subregion of each region. There are 150 S's
+    in Region 1 and 10 S's in region 2, therefore, we have:
+
+    +----------+-------------+--------------------------------+
+    | Region   | Subregion   | Fraction of all As in Region   |
+    +==========+=============+================================+
+    | Region 1 | Subregion A | 100 / 150 = 0.6667             |
+    +----------+-------------+--------------------------------+
+    | Region 1 | Subregion B | 50 / 150 = 0.3333              |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion C | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion D | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion E | 10 / 10 = 1.0000               |
+    +----------+-------------+--------------------------------+
+
+    Finally, for each subregion, we multiply these together and
+    add them up the values for the subregions in each region.
+    For Region 1, we get
+
+    .. math::
+        (0.6667 * 1.0) + (0.3333 * 0.5) = 0.8333.
+
+    For Region 2, we get
+
+    .. math::
+        0.0 * 0.0 + 0.0 * 0.0 + 1.000 * 0.1 = 0.1.
+
+    Note that the implentation may not do this math exactly as
+    specified here, but it will do something equivalent.
+    """
+    if by is None:
+        df_communities["REGION"] = "Region " + df_communities.index.astype(str)
+        by = "REGION"
+    if over is None:
+        df_communities["SUBREGION"] = "Subregion " + df_communities.index.astype(str)
+        over = "SUBREGION"
+    else:
+        subregion_groups = df_communities.groupby(over)
+        for _, group in subregion_groups:
```

Comment:
I think the check you are doing here is not necessary. It might be the case that `Region 1` and `Region 2` both have subregions called `A`, `B`, and `C`. This commonly happens in census data. For example, each state has a code, e.g. California is `"06"` and New Jersey is `"33"`. But within those, the counties are given codes, `"001"`, `"002"`, `"003"`, and so on. So if we used this API with `by='STATE'` and `over='COUNTY'` on data that had all the counties in two or more states, I think it could trigger this error even though it is a valid input.